File: Bot.cpp
#include "Bot.h"
#include "Position/Position.h"
#include "ExchangeAPI/ExchangeWrapper.h"

#include <string>
#include <utility> 
#include <fstream>
#include <iostream>
#include <ctime>

Bot::Bot(std::string exchange1, std::string exchange2, double quantity){
	ExchangeWrapper wrapper;
	ex1_ = wrapper(exchange1, quantity); 
	ex2_ = wrapper(exchange2, quantity); 
	quantity_ = quantity;
}

Bot::~Bot(){
	delete ex1_; ex1_ = nullptr;
	delete ex2_; ex2_ = nullptr;
}

void Bot::startBot(long endTime){

    while(std::time(nullptr) < endTime){
		std::pair<Position, bool> potentialArb = tryArbitrage();
		if (std::get<1>(potentialArb) == true){
			writeToTextFile(std::get<0>(potentialArb));
		} 
    }
}

void Bot::startBot(int numTrades){

	int totalTrades = 0;
    while(totalTrades < numTrades){
		std::pair<Position, bool> potentialArb = tryArbitrage();
		if (std::get<1>(potentialArb) == true){
			writeToTextFile(std::get<0>(potentialArb));
			totalTrades++;   
		} 
    }
}

std::pair<Position, bool> Bot::tryArbitrage(){
		
	double e1Spot = ex1_->getSpotPrice();
	double e2Spot = ex2_->getSpotPrice();

	if(isProfitableSpread(e1Spot, e2Spot)){
		Position tradeData = executeArbitrage(e1Spot, e2Spot);
		return std::pair<Position,bool>(tradeData, true);
	}
	else {
		return std::pair<Position,bool>(Position(), false);
	}
}

void Bot::writeToTextFile(Position p){
	std::string pos = "";
	pos += (std::to_string(p.q) + ",");
	pos += (std::to_string(p.sP) + ",");
	pos += (std::to_string(p.lP) + ",");
	pos += ((std::to_string(p.completionTime) + "\n"));

	std::cin >> pos;
    std::ofstream out("closedPositions.txt");
    out << pos;
}

Position Bot::executeArbitrage(double spot1, double spot2){
	//TODO: IMPLEMENT ARBITRAGE
	ex1_->goLong(); ex2_->goShort();
	//TODO: VERIFY FUNDS ARE SETTLED
	Position tradeData(quantity_, spot1, spot2);
	tradeData.completionTime = std::time(nullptr);
	tradeData.isComplete = true;
	return tradeData;
}

bool Bot::isProfitableSpread(double spot1, double spot2){

	//TODO: Include additional parameters beyond just ROI
	double e1Commission = ex1_->getCommission();
	double e2Commission = ex2_->getCommission();
	double spread = spot1-spot2;
	double adjSpread = spread-(spread*(e1Commission+e2Commission));

	if (adjSpread > 0){
		return true;
	}
	return false;
}

File: PyArbitrage/Bot.py

class Bot(object):

	def __init__(self, quantity, e1Keys, e2Keys):
		from GDAX_API import GDAX
		from CEX_API import CEX
		self.gdaxWrapper = GDAX(quantity, *e1Keys)
		self.cexWrapper = CEX(quantity, *e2Keys)
		self.generateTextFile()
		self.quantity = quantity

	#### USER-END FUNCTIONS ####
	def start(self, endTime=None, numTrades=None):

		if (endTime != None):
			import time
			while(time.time() < endTime):
				potentialArb = self.tryArbitrage()
				if (potentialArb[1]):
					self.writeToTextFile(potentialArb[0])

		if (numTrades != None):
			totalTrades = 0 
			while (totalTrades < numTrades):
				potentialArb = self.tryArbitrage()
				if (potentialArb[1]):
					self.writeToTextFile(potentialArb[0])
					totalTrades += 1

	#### ARBITRAGE FUNCTIONS ####
	def tryArbitrage(self):
		longSpot = self.bfxWrapper.getSpotPrice()
		shortSpot = self.cexWrapper.getSpotPrice()

		if (self.isProfitableSpread(longSpot, shortSpot)):
			return (executeArbitrage(longSpot, shortSpot), True)
		else:
			import Position as pos
			return (pos.Position(), False)

	def isProfitableSpread(self, longSpot, shortSpot):
		#TODO: Include additional parameters beyond just ROI
		lCommission = self.bfxWrapper.getCommission()
		sCommission = self.cexWrapper.getCommission()
		spread = shortSpot-longSpot;
		adjSpread = spread-(spread*(lCommission+sCommission))

		if (adjSpread > 0):
			return true
		else:
			return false

	def executeArbitrage(self, longSpot, shortSpot):
		#TODO: IMPLEMENT ARBITRAGE
		self.bfxWrapper.goLong()
		self.cexWrapper.goShort()

		#TODO: VERIFY FUNDS ARE SETTLED
		import Position as pos
		tradeData = pos.Position(self.quantity, longSpot, shortSpot)
		tradeData.setCompleteParams()
		return tradeData

	##### LOGGING FUNCTIONS #####
	def generateTextFile(self):
		import Position as pos
		headers = list(pos.Position().toDict().keys())
		commaDelim = ",".join(headers)

		f = open('trades.txt', 'w+')
		f.write(commaDelim + "\n")
		f.close()

	def writeToTextFile(self, pos): 

		if (pos.isComplete == False):
			import time
			pos.completionTime = time.time()
		
		vals = [str(elem) for elem in pos.toDict().values()]
		posStr = ",".join(vals)
		f = open('trades.txt', 'a')
		f.write(posStr + "\n")
		f.close()
